doc/developer/design/20240311_adapter_catalog_listen.md

```diff
@@ -0,0 +1,178 @@
+by other coordinators/adapters. In order to facilitate this in-memory catalog state updates,
+builtin table updates, and controller commands should be created in response to a change to the
+durable catalog state.
+
+## Success Criteria
+
+All `sequence_X` methods that correspond to DDL should be structured in the following way:
+
+1. Keep doing the same thing before the catalog transaction.
+2. Perform catalog the catalog transaction which only updates the durable catalog state.
+3. Retrieve all changes to the durable catalog.
+4. Update in-memory catalog state based on the changes.
+5. Generate builtin table updates and controller commands from the changes.
+
+`Catalog::open` should be structured in a way that all updates to in-memory state, builtin table
+updates, and controller commands are generated directly from a list of updates from the durable
+catalog.
+
+## Out of Scope
+
+- A catalog subscribe API.
+- Multi-subscriber catalog.
+
+## Solution Proposal
+
+### Data Structures
+
+```Rust
+struct StateUpdate {
+    kind: StateUpdateKind,
+    diff: Diff,
+}
+
+enum StateUpdateKind {
+    AuditLog(mz_audit_log::VersionedEvent),
+    Cluster(mz_catalog::durable::objects::Cluster),
+    ClusterReplica(mz_catalog::durable::objects::ClusterReplica),
+    Comment(mz_catalog::durable::objects::Comment),
+    Database(mz_catalog::durable::objects::Database),
+    DefaultPrivilege(mz_catalog::durable::objects::DefaultPrivilege),
+    IntrospectionSourceIndex(mz_catalog::durable::objects::IntrospectionSourceIndex),
+    Item(mz_catalog::durable::objects::Item),
+    Role(mz_catalog::durable::objects::Role),
+    Schema(mz_catalog::durable::objects::Schema),
+    StorageUsage(mz_audit_log::VersionedStorageUsage),
+    SystemConfiguration(mz_catalog::durable::objects::SystemConfiguration),
+    SystemObjectMapping(mz_catalog::durable::objects::SystemObjectMapping),
+    SystemPrivilege(MzAclItem),
+}
+
+/// Enum of all commands that the Coordinator can send to the controller.
+enum ControllerCommand {
+    // ...
+}
+```
+
+### Durable Catalog API
+
+```Rust
+trait ReadOnlyDurableCatalogState {
+    // ...
+
+    /// Fetches and returns all updates to the durable catalog state that have not yet been
+    /// consumed.
+    async fn sync(&mut self) -> Vec<StateUpdate>;
+
+    // ...
+}
+```
+
+### In-memory Catalog API
+
+```Rust
+impl Catalog {
+    /// Fetches and applies all updates to the durable catalog state that have not yet been
+    /// consumed.
+    async fn sync(&mut self) -> (Vec<BuiltinTableUpdate>, Vec<ControllerCommand>) {
+        let updates = self.storage().await.sync().await;
+        self.apply_updates(updates);
+    }
+
+    /// Update in-memory catalog state and generate builtin table updates and controller commands. 
+    fn apply_updates(&mut self, updates: Vec<StateUpdate>) -> (Vec<BuiltinTableUpdate>, Vec<ControllerCommand>) {
```

Comment:
Ah cool, and you don't have to worry about idempotence because "Multi-subscriber catalog" is out of scope, nice!